### ISSUE TITLE
About 'shift-pooling operator'

### ISSUE THREAD

**Issue by SlingHe** [dingmyu/D4LCN#1]
Hi dingmyu, you mention a shift-pooling operator in your paper. However, I didn't find any relevant implementation in this repo. Any suggestions or follow-ups? Thanks :)

**Comment by dingmyu** [comment]
Hi SlingHe, Thanks for your interest.

The shift-pooling is implemented in the function ``dynamic_local_filtering'' at https://github.com/dingmyu/D4LCN/blob/master/models/resnet_dilate.py#L13. Note that here we provide a simplified and faster version with comparable performance.

It can be implemented simply, for example:

    y = torch.cat((x[:, -1:, :, :], x[:, :-1, :, :]), dim=1)
    z = torch.cat((x[:, -2:, :, :], x[:, :-2, :, :]), dim=1) 
    x = (x + y + z)/3

If you modify the parameter $n_f$ in the shift-pooling w or w/o BN/RELU layers, a better result than it in the paper may be achieved. Please feel free to contact me if you need any further information.

**Comment by dingmyu** [comment]
Feel free to reopen it if you have any further questions.

**Comment by dannyhung1128** [comment]
Hi, what do you mean a better result than paper may be achieved? Did you try different settings of $n_f$? Can you please provide some experimental results or findings with different $n_f$? Thanks!

**Comment by dingmyu** [comment]
> Hi, what do you mean a better result than paper may be achieved? Did you try different settings of $n_f$? Can you please provide some experimental results or findings with different $n_f$? Thanks!

Hi Danny,

I did this experiment and found that $n_f = 3$ is the best (in the paper), $n_f = 2 or 4$ is slightly worse than $n_f = 3$.

However, when I clean up this code, I found that simplifying the shift-pooling and modifying BN / relu / dropout sometimes lead to better results. For example, $n_f = 2$, remove BN, increase the dropout rate. I think it is because of the overfitting problem in the KITTI 3D dataset.

Thank you.
